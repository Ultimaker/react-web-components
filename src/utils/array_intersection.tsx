// returns an array containing the common elements of two given arrays
export function arrayIntersection(firstArray: any[], secondArray: any[]): any[] {
    if (Array.isArray(firstArray) && Array.isArray(secondArray)) {
        return firstArray.filter((firstArrayElement) => (
            secondArray.includes(firstArrayElement)));
    }
    return [];
}

export default arrayIntersection;
