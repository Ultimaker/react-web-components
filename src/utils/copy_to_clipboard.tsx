// Copyright (c) 2018 Ultimaker B.V.

/**
 * Copies a string to the user's clipboard.
 * @param value - The string to be copied.
 */
export function copyToClipboard(value: string){
    let selectionTextarea = document.createElement('textarea');
    selectionTextarea.style.position = 'fixed';
    selectionTextarea.style.left = '0';
    selectionTextarea.style.top = '0';
    selectionTextarea.style.opacity = '0';
    selectionTextarea.value = value;
    document.body.appendChild(selectionTextarea);
    selectionTextarea.focus();
    selectionTextarea.select();
    document.execCommand('copy');
    document.body.removeChild(selectionTextarea);
}
  
export default copyToClipboard;
