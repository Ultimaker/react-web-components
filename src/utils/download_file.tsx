// Copyright (c) 2018 Ultimaker B.V.

/**
 * Allows the user to download a file with the given contents.
 * @param fileName - The default name of the file.
 * @param fileContent - The content of the file.
 * @param contentType - The content type, including the charset.
 * Defaults to "text/plain;charset=utf-8".
 */
function downloadFile(fileName: string, fileContent: string, contentType?: string) {
    const blob = new Blob([fileContent], { type: contentType || 'text/plain;charset=utf-8' });
    const element = document.createElement('a');
    element.href = URL.createObjectURL(blob);
    element.download = fileName;
    element.click();
}

export default downloadFile;
