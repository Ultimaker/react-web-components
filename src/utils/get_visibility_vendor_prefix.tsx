function getVisibilityVendorPrefix(): string {
    const prefixes = ['webkit', 'moz', 'ms', 'o'];

    // if 'hidden' is natively supported return nothing
    if ('hidden' in document) return '';

    // otherwise loop over all the known prefixes until we find one
    for (var i = 0; i < prefixes.length; i++) {
        if ((prefixes[i] + 'Hidden') in document)
            return prefixes[i];
    }

    // otherwise it's not supported
    return null;
}

export default getVisibilityVendorPrefix;
