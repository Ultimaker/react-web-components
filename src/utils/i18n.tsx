// Copyright (c) 2018 Ultimaker B.V.
import Gettext = require('node-gettext');
import { po } from 'gettext-parser'
import * as React from 'react'

// type for available languages
export type Languages = 'en-US' | 'nl-NL'

// interface for translation items that are passed into initialize
export interface TranslationListItem {
	name: Languages
	source: any
}

/** The translated text may be a string or a list of strings and elements (when interpolate is used) **/
export type TranslatedText = string | (string | JSX.Element)[]

/**
 * Assorted methods for translations.
 */
export class I18n {

	private static _gt = new Gettext({ debug: false })
	private static _defaultLanguage = 'en-DEV'
	private static _supportedLanguages: Languages[] = ['en-US', 'nl-NL']

	/**
	 * Initialize the Gettext context.
	 */
	public static async initialize(translations: TranslationListItem[]) {

		// load the translation files
		translations.forEach(async translation => await this._loadTranslation(translation.name, translation.source))

		// set the default locale based on the browser settings
		I18n._gt.setLocale(I18n.getLocale())
	}

	/**
	 * Get the locale determined by the browser.
	 */
	public static getLocale (): string {

		// determine the first supported language using navigator.languages
		if (navigator.languages !== undefined) {
			return navigator.languages.find(I18n.isLanguageSupported)
		}

		// determine the language using navigator.language (fallback 1)
		if (navigator.language !== undefined) {
			return [navigator.language].find(I18n.isLanguageSupported)
		}

		// determine the language using navigator.browserLanguage (fallback 2)
		if (navigator.hasOwnProperty('browserLanguage')) {
			return [navigator['browserLanguage']].find(I18n.isLanguageSupported)
		}

		// return the default language
		return I18n._defaultLanguage
	}

	/**
	 * Check if the given language is supported.
	 * @param language The language to check, for example 'en'.
	 */
	public static isLanguageSupported (language: Languages): boolean {
		return I18n._supportedLanguages.indexOf(language) > -1
	}

	/**
	 * Get a translated version of the given text.
	 * @param context Context markers for this text to help translators.
	 * @param text The text to translate or fallback to.
	 */
	public static translate (context: string, text: string): string {
		return I18n._gt.pgettext(context, text)
	}

	/**
	 * Get a translated and formatted version of the given text.
	 * @param context Context markers for this text to help translators.
	 * @param text The text to translate and format.
	 * @param parameters The parameters to insert in the text.
	 */
	public static format(context: string, text: string, parameters: {[key: string]: any}): TranslatedText {
		return I18n.interpolate(this.translate(context, text), parameters)
	}

	/**
	 * Replace all passed parameters in a text.
	 * @param text The text to interpolate.
	 * @param parameters The parameters to insert in the text.
	 */
	public static interpolate(text: string, parameters: {[key: string]: any}): TranslatedText {
		return text.split(/%{(\w+)}/g).map((part, i) => part &&
			<span key={i}>{parameters[part] || part}</span>
		)
	}

	/**
	 * Loads a language source file async.
	 * @param language - The language we're loading.
	 * @param url - The url of the language source file.
	 */
	private static async _loadTranslation (language: Languages, url: string): Promise<void> {
		await fetch(url)
			.then(response => response.text())
			.then(text => I18n._parse(text))
			.then(translations => I18n._gt.addTranslations(language, 'messages', translations))
	}

	/**
	 * Parse a translations source file into an object that gettext can read.
	 * @param source Translations source file.
	 */
	private static _parse (source: string): any {
		return po.parse(source)
	}
}
