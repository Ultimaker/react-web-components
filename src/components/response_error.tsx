// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';

// utils
import { I18n } from '../utils/i18n';

// components
import Tile from './tile';


/**
 * Gets a translated message for the given error.
 * @param error - The error object to be translated.
 * @param errorMessageTemplates - The error translation templates.
 * @param fieldTranslations - The field translation templates.
 */
export const getTranslatedError = (
    error: ResponseErrorObject,
    errorMessageTemplates: { [key: string]: () => string },
    fieldTranslations: { [key: string]: () => string },
): string => {
    const params = { ...error };

    // check if we need to translate any fields
    if (params.meta) {
        Object.keys(error.meta).forEach((key) => {
            if (key === 'field_name' && fieldTranslations[params[key]]) {
                // each field translation is a function that returns a string
                params[key] = fieldTranslations[params[key]]();
            } else {
                params[key] = error.meta[key];
            }
        });
    }

    // generate the error message
    // each error message template is a function that returns a string
    const errorMessage = errorMessageTemplates[error.code]
        ? errorMessageTemplates[error.code]()
        : `${error.title} (${error.code})`;

    return I18n.interpolate(errorMessage, params);
};


export interface ResponseErrorObject {
    id: string;
    code: string;
    http_status: string; // eslint-disable-line camelcase
    title: string;
    detail?: string;
    meta?: { [key: string]: any };
}

export interface ResponseErrorProps {
    errors: ResponseErrorObject[];
    errorMessageTemplates?: { [key: string]: () => string };
    fieldNames?: { [key: string]: () => string };

    popupHeaderText: string;            // Deprecated
    popupBodyText: string;              // Deprecated
    popupDownloadButtonText: string;    // Deprecated
    popupCancelButtonText: string;      // Deprecated
}

export class ResponseError extends React.Component<ResponseErrorProps, {}> {

    render(): JSX.Element {
        const {
            errors, errorMessageTemplates, fieldNames
        } = this.props;

        return (
            <div className="response-error">
                <Tile align="center" alert>
                    {errors.map((error) => (
                        <div className="response-error__msg" key={error.id}>
                            {getTranslatedError(error, errorMessageTemplates, fieldNames)}
                        </div>
                    ))}
                </Tile>
            </div>
        );
    }
}

export default ResponseError;
