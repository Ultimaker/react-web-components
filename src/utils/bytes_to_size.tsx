/* http://stackoverflow.com/questions/15900485/correct-way-to-convert-size-in-bytes-to-kb-mb-gb-in-javascript */

import {I18n} from './i18n'

const sizes = [
  I18n.translate('bytes to size', 'Bytes'),
  I18n.translate('bytes to size', 'KB'),
  I18n.translate('bytes to size', 'MB'),
  I18n.translate('bytes to size', 'GB'),
  I18n.translate('bytes to size', 'TB'),
];

function bytesToSize(bytes: number): string {
  if (bytes === 0) return '0 Byte';
  const i = Math.floor(Math.log(bytes) / Math.log(1024));
  return (Math.round(100 * bytes / Math.pow(1024, i)) / 100).toString() + sizes[i];
}

export default bytesToSize;
