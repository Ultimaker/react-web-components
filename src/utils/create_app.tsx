// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { AppContainer } from 'react-hot-loader';
import { BrowserRouter } from 'react-router-dom';

import { I18n } from '../utils/i18n';

declare const module: any;

/**
 * Create a standard app.
 * @param domId The DOM element on which to render the app.
 * @param appComponent The main application component to generate the app with.
 * @param localesPath The directory location of the translation files.
 */
export function createApp(domId: string, AppComponent, localesPath: string) {
    
    // Set locale
    let locale = I18n.getLocale();
    console.info('locale', locale);
    I18n.setMomentLocale(locale);

    I18n.load({
        path: localesPath,
        locale: locale,
        complete: () => {
            _renderApp(domId, AppComponent);
        }
    });

    // Hot Module Replacement API
    if (module.hot) {
        module.hot.accept(() => {
            _renderApp(domId, AppComponent);
        })
    }
}

/**
 * Render a standard app.
 * @param domId The DOM element on which to render the app.
 * @param appComponent The root app component.
 */
function _renderApp(domId: string, AppComponent) {
    ReactDOM.render(
        <AppContainer>
            <BrowserRouter>
                <AppComponent />
            </BrowserRouter>
        </AppContainer>, document.getElementById(domId)
    )
}
