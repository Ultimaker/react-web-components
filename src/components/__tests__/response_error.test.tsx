// Copyright (c) 2018 Ultimaker B.V.
import * as React from 'react';
import { shallow } from 'enzyme';

// utils
import { I18n } from '../../utils/i18n';

// component
import { ResponseError, ResponseErrorProps } from '../response_error';

const ERROR_CODE_TO_I18NC_TEMPLATE = {
    fieldError: () => I18n.translate('error message', '%{field_name} was invalid: %{title}.'),
};

const FIELD_NAME_TO_I18NC = {
    username: () => I18n.translate('form field username', 'Username'),
};

describe('The ResponseError component', () => {
    let props: ResponseErrorProps;
    let wrapper;

    beforeEach(() => {
        props = {
            errorMessageTemplates: ERROR_CODE_TO_I18NC_TEMPLATE,
            fieldNames: FIELD_NAME_TO_I18NC,
            popupHeaderText: 'Something went wrong at our end :(',
            popupBodyText: 'Please describe here what you were doing that caused the error to happen. Then download the report and attach this in an email to your Ultimaker reseller.',
            popupDownloadButtonText: 'Download',
            popupCancelButtonText: 'Cancel',
            errors: [{
                id: '12345',
                code: 'randomError',
                http_status: '403',
                title: 'The server crashed hard',
            }],
        };
        wrapper = shallow(<ResponseError {...props} />);
    });

    it('should render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render translated error message', () => {
        wrapper.setProps({
            errors: [{
                id: '12345',
                code: 'fieldError',
                title: 'Field was invalid',
                detail: 'The field has an invalid value',
                meta: {
                    field_name: 'username',
                },
            }],
        });
        expect(wrapper.find('.response-error__msg')).toHaveLength(1);
        expect(wrapper.find('.response-error__msg').text()).toBe('username was invalid: Field was invalid.');
    });
});
