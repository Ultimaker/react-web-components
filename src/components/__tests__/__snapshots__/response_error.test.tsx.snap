// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ResponseError component should render 1`] = `
<div
  className="response-error"
>
  <Tile
    alert={true}
    align="center"
    appearance="shadow"
    padding="md"
  >
    <div
      className="response-error__msg"
      key="12345"
    >
      There was a problem on the server: The server crashed hard (code randomError)
    </div>
  </Tile>
</div>
`;
